src/main/java/org/hydev/mcpm/client/Downloader.java

```diff
@@ -40,8 +83,45 @@ public void downloadFile(String url, File to)
      */
     public void downloadFiles(Map<String, File> urls, boolean progress, int threads)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        Thread[] threadss = new Thread[threads];
+        final int UrlPerThread = urls.size() / threads;
+        final int remainingUrl = urls.size() % threads;
+
+        for (int t = 0; t < threads; t++) {
+            final int thread = t;
+            threadss[t] = new Thread() {
+                @Override
+                public void run() {
+                    runThread(urls, threads, thread, UrlPerThread, remainingUrl);
+                }
+            };
+        }
+
+        for (Thread t: threadss) {
+            t.start();
+        }
+    }
+    private void runThread(Map<String, File> urls, int threads, int thread, int UrlPerThread, int remainingUrl) {
+        // Store all urls into a list
+        List<String> files = new ArrayList<>();
+        for (Map.Entry<String, File> url:urls.entrySet()) {
+            files.add(url.getKey());
+        }
+        // Store urls per thread
+        List<String> inFiles = new ArrayList<>();
+
+        for (int i=thread*UrlPerThread; i < (thread + 1) * UrlPerThread; i++) {
+            inFiles.add(files.get(i));
+        }
+        if (thread == threads-1 && remainingUrl > 0) {
+            for (int j = files.size() - remainingUrl; j < files.size(); j++) {
+                inFiles.add(files.get(j));
+            }
+        }
+
+        for (String url:inFiles) {
+            downloadFile(url, urls.get(url));
```

Comment:
It looks like you have multiple ProgressBars for every thread.
 - Can you create one big `ProgressBar` for the whole operation and pass a `ProgressRow` down to `downloadFile`?
  - Are there any issues with multiple `ProgressBar`s working at the same time? Is `ProgressBar`s thread safe? 
  - You might need to lock `ProgressBar`s with a mutex or a `synchronized (bar) { ... }` block.
  
 I'll ask I guess and come back to this comment.